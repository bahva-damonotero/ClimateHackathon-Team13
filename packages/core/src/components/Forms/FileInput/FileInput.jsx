import React, { useEffect, useRef } from 'react';
import PropTypes from 'prop-types';
import fileInput from '@uswds/uswds/js/usa-file-input';

import { FormError, Label } from '..';
import { Hint } from '../..';

export default function FileInput({
  accept,
  children,
  disabled,
  error,
  hint,
  id,
  multiple,
  onChange,
  value,
  wrongFileTypeErrorMessage
}) {
  const containerRef = useRef();

  useEffect(() => {
    const containerEl = containerRef.current;
    if (!containerEl) return;

    // call USWDS's `on` function to initialize the component
    fileInput.on(containerEl);

    // on dismount, call USWDS's `off` function to cleanup the component
    return () => {
      fileInput.off(containerEl);
    };
  }, []);

  // call USWDS's `enable` or `disable` function when the disabled prop changes
  useEffect(() => {
    if (!containerRef.current) return;
    const inputEl = containerRef.current.querySelector(`#${id}`);

    const disabledState = inputEl.disabled;
    if (disabled === disabledState) return;

    // if disabled doesnt already have the right state
    if (disabled) fileInput.disable(inputEl);
    else fileInput.enable(inputEl);
  }, [disabled]);

  const hintId = `${id}--hint`;
  return (
    <div className="bah-uswds-file-input" ref={containerRef}>
      <Label target={id}>{children}</Label>

      <FormError error={error} />

      {hint ? <Hint id={hintId}>{hint}</Hint> : <></>}

      <div className="bah-uswds-file-input--wrapper">
        <input
          accept={accept}
          aria-describedby={hint ? hintId : undefined}
          className="usa-file-input"
          id={id}
          multiple={multiple}
          name={id}
          onChange={onChange}
          type="file"
          value={value}
          data-errormessage={wrongFileTypeErrorMessage}
        />
      </div>
    </div>
  );
}

FileInput.propTypes = {
  /**
   * String defining the file types the input should accept.
   * See here: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/file#attr-accept
   */
  accept: PropTypes.string,
  /** Label for input */
  children: PropTypes.node.isRequired,
  /** Should the file input be disabled? */
  disabled: PropTypes.bool,
  /** Error message to display */
  error: PropTypes.string,
  /** Hint to be shown for the file input */
  hint: PropTypes.node,
  /** ID of the input */
  id: PropTypes.string.isRequired,
  /** Should the file input accept multiple files? */
  multiple: PropTypes.bool,
  /** Function called when the file selection changes */
  onChange: PropTypes.func,
  /** Used to set the value for the file input */
  value: PropTypes.object,
  /** Used to override the error message shown when an invalid file type is selected */
  wrongFileTypeErrorMessage: PropTypes.string
};

FileInput.defaultProps = {
  disabled: false,
  multiple: false
};
