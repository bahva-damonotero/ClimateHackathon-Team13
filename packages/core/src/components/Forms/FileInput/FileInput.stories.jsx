import React from 'react';
import { action } from '@storybook/addon-actions';

import FileInput from '.';
import { Form } from '..';

export default {
  title: 'Forms/FileInput',
  component: FileInput
};

const Template = (args) => (
  <Form>
    <FileInput {...args} />
  </Form>
);

const defaultArgs = {
  children: 'Label',
  id: 'fileinput-example',
  onChange: action('onChange')
};

export const def = Template.bind({});
def.storyName = 'Default';
def.args = {
  ...defaultArgs
};

export const fileTypesRestriction = Template.bind({});
fileTypesRestriction.args = {
  ...defaultArgs,
  accept: 'application/pdf,text/plain',
  hint: 'Select PDF or TXT files'
};

export const multipleFiles = Template.bind({});
multipleFiles.args = {
  ...defaultArgs,
  multiple: true
};
