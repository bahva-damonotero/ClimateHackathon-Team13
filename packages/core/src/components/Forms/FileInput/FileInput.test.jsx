import React from 'react';
import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import fileInput from '@uswds/uswds/js/usa-file-input';

import FileInput from '.';

describe('FileInput', () => {
  const id = 'file-input-example';
  const label = 'Select a file to upload';

  it('should call `fileInput.on()` to initialize the component', () => {
    const onSpy = jest.spyOn(fileInput, 'on');

    render(<FileInput id={id}>{label}</FileInput>);

    expect(onSpy).toHaveBeenCalled();
  });

  it('changing the `disabled` prop to true should call `fileInput.disable()`', () => {
    const disableSpy = jest.spyOn(fileInput, 'disable');

    const { rerender } = render(<FileInput id={id}>{label}</FileInput>);

    rerender(
      <FileInput disabled id={id}>
        {label}
      </FileInput>
    );

    expect(disableSpy).toHaveBeenCalled();
  });

  it('changing the `disabled` prop to false should call `fileInput.enable()`', () => {
    const enableSpy = jest.spyOn(fileInput, 'enable');

    const { rerender } = render(
      <FileInput disabled id={id}>
        {label}
      </FileInput>
    );

    rerender(<FileInput id={id}>{label}</FileInput>);

    expect(enableSpy).toHaveBeenCalled();
  });

  it('should match snapshot', () => {
    const tree = renderer
      .create(
        <FileInput
          accept="application/pdf,text/plain"
          hint="Select PDF or TXT files"
          id={id}
        >
          {label}
        </FileInput>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
