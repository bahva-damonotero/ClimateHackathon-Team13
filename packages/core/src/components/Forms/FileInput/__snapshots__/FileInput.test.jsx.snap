// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileInput should match snapshot 1`] = `
<div
  className="bah-uswds-file-input"
>
  <label
    className="usa-label"
    htmlFor="file-input-example"
  >
    Select a file to upload
  </label>
  <span
    className="usa-hint"
    id="file-input-example--hint"
  >
    Select PDF or TXT files
  </span>
  <div
    className="bah-uswds-file-input--wrapper"
  >
    <input
      accept="application/pdf,text/plain"
      aria-describedby="file-input-example--hint"
      className="usa-file-input"
      id="file-input-example"
      multiple={false}
      name="file-input-example"
      type="file"
    />
  </div>
</div>
`;
